### PR TITLE
`ListUsers` intersection with wildcards

### PR DESCRIPTION
**Pull request by willvedd**
## Description
The initial implementation of intersection (#1467) in `ListUsers` missed an important case during testing, public typed wildcard tuples. Wildcards require special consideration when it comes to intersection. 

The initial implementation of intersection relied on a counter per user object, and if that count equaled the number of `and` clauses, then the user would be returned on the main results channel. However the presence of wildcards in a clause needs to be considered; you do not want to double-count users as shown in this example below:

```
model
  schema 1.1
type user
type document
  relations
    define public_1: [user:*,user]
    define public_2: [user:*,user]
    define is_public: public_1 and public_2
```
```
"document:1", "public_1", "user:maria"
"document:1", "public_1", "user:*",
```
```
ListUsers("document:1", user_filter: ["user"], relation: "is_public") // should return empty list
```

The solution provided in this PR is to individually evaluate a separate channel per clause to be able to consider the wildcard case.

## References
- Related PR: #1467 
- Related comment: https://github.com/openfga/openfga/pull/1467#discussion_r1533089380
- Parent PR: #1428 

## Review Checklist
- [x] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [x] The correct base branch is being used, if not `main`
- [x] I have added tests to validate that the change in functionality is working as expected

### CODE REVIEW

File: pkg/server/commands/listusers/list_users_rpc_test.go
Comment (miparnisari) [inline pkg/server/commands/listusers/list_users_rpc_test.go:1651]: these two tuples aren't allowed by the model you defined

File: pkg/server/commands/listusers/list_users_rpc_test.go
Comment (miparnisari) [inline pkg/server/commands/listusers/list_users_rpc_test.go:1589]: these two tuples aren't allowed by the model you defined

File: pkg/server/commands/listusers/list_users_rpc_test.go
Comment (miparnisari) [inline pkg/server/commands/listusers/list_users_rpc_test.go:1499]: nit: indent

File: pkg/server/commands/listusers/list_users_rpc_test.go
Comment (willvedd) [reply pkg/server/commands/listusers/list_users_rpc_test.go:1589]: That's exactly what I'm trying to test. Or is it safe to assume that this is disallowed by the datastore?

Comment (miparnisari) [reply pkg/server/commands/listusers/list_users_rpc_test.go:1595]: The datastore doesn't check for correctness, the API has to do that. So you have make sure that the API filters out invalid tuples with respect to the model being used. There are two approaches for this:

1. Use a filtered tuple iterator `filteredIter := storage.NewFilteredTupleKeyIterator(storage.NewTupleKeyIteratorFromTupleIterator(tupleIter), validation.FilterInvalidTuples(typesys))`
2. Use the graph of the model to pre-filter what tuples to fetch (like we do in ReverseExpand)

File: pkg/server/commands/listusers/list_users_rpc.go
Comment (miparnisari) [inline pkg/server/commands/listusers/list_users_rpc.go:385]: I'm not a huge fan of this approach. It probably works but it looks complex, and you'll have to repeat it for exclusion. 

I feel like we're making our lives harder by not re-using the ReverseExpansion approach of using the graph traversal 🤔

File: pkg/server/commands/listusers/list_users_rpc_test.go
Comment (willvedd) [reply pkg/server/commands/listusers/list_users_rpc_test.go:1595]: We currently employ strategy 1. This test is ensuring that it works correctly. I have a strong preference to maintain this test case because it is easily forgotten and can wreak havoc on our system.

File: pkg/server/commands/listusers/list_users_rpc_test.go
Comment (willvedd) [reply pkg/server/commands/listusers/list_users_rpc_test.go:1657]: Discussed in similar comment below.

File: pkg/server/commands/listusers/list_users_rpc.go
Comment (willvedd) [reply pkg/server/commands/listusers/list_users_rpc.go:385]: > I'm not a huge fan of this approach. It probably works but it looks complex, and you'll have to repeat it for exclusion.

I'm not in love either but it works and it makes sense why it works, although this is hugely subjective. 

Fortunately we don't need to repeat for exclusion. The complexity stems from supporting any number of clauses whereas exclusion always deal with two clauses.

> I feel like we're making our lives harder by not re-using the ReverseExpansion approach of using the graph traversal 

I still intend on having us evaluate wether reverse expand and/or some other algorithm is more appropriate, but for now I'm mostly concerned on getting an initial feature-complete version implemented.

File: pkg/server/commands/listusers/list_users_rpc.go
Comment (jon-whit) [inline pkg/server/commands/listusers/list_users_rpc.go:326]:
```suggestion
	childOperands := rewrite.Intersection.GetChild()
```

File: pkg/server/commands/listusers/list_users_rpc.go
Comment (jon-whit) [inline pkg/server/commands/listusers/list_users_rpc.go:355]:
```suggestion
		go func(foundUsersChan chan *openfgav1.User) {
```

File: pkg/server/commands/listusers/list_users_rpc.go
Comment (jon-whit) [inline pkg/server/commands/listusers/list_users_rpc.go:357]:
```suggestion
			for foundUser := range foundUsersChan {
```

File: pkg/server/commands/listusers/list_users_rpc.go
Comment (jon-whit) [inline pkg/server/commands/listusers/list_users_rpc.go:324]: This should be a helper in the `pkg/tuple` package.

```go
package tuple
...
func TypedPublicWildcard(objectType string) string {
    return fmt.Sprintf("%s:%s", objectType, Wildcard)
```

File: pkg/server/commands/listusers/list_users_rpc.go
Comment (jon-whit) [inline pkg/server/commands/listusers/list_users_rpc.go:374]: Please `defer wg.Done()` at the top of this function.

File: pkg/server/commands/listusers/list_users_rpc.go
Comment (jon-whit) [reply pkg/server/commands/listusers/list_users_rpc.go:389]: I don’t have any strong opposition to the algorithm at this time. It seems reasonable for the current API invariants. It would get much uglier if we supported more than 1 user filter in the near future. That would be my main concern with the approach, because that multiply the number of typed wildcards and user objects we'd have to track independently as well.

File: pkg/server/commands/listusers/list_users_rpc_test.go
Comment (jon-whit) [inline pkg/server/commands/listusers/list_users_rpc_test.go:1930]: Could we add another test that walks some userset expansion stages with wildcards?

```
model
  schema 1.1

type user

type group
  relations
    define member: [user:*, user]

type document
  relations
    define group: [group]
    define viewer: [group#member] and member from group
    define can_view: viewer
```
- document:1#viewer@group:y#member
- document:1#group@group:x
- group:x#member@user:*
- group:y#member@user:jon

`ListUsers(document:1#can_view, user_filters=[{"type": "user"}])` should return ["user:jon"]

File: pkg/server/commands/listusers/list_users_rpc_test.go
Comment (miparnisari) [reply pkg/server/commands/listusers/list_users_rpc_test.go:1837]: Ok, then can you please leave a comment next to each "invalid" tuple ?

File: pkg/server/commands/listusers/list_users_rpc_test.go
Comment (miparnisari) [reply pkg/server/commands/listusers/list_users_rpc_test.go:1899]: Please leave a comment next to each invalid tuple.

File: pkg/server/commands/listusers/list_users_rpc.go
Comment (miparnisari) [inline pkg/server/commands/listusers/list_users_rpc.go:324]: nit: please move this closer to where it's used (line 360)

File: pkg/server/commands/listusers/list_users_rpc.go
Comment (miparnisari) [inline pkg/server/commands/listusers/list_users_rpc.go:352]:
```suggestion
	wildcardCount := atomic.Uint32{} // guarded by mu
```

File: pkg/server/commands/listusers/list_users_rpc.go
Comment (miparnisari) [inline pkg/server/commands/listusers/list_users_rpc.go:353]:
```suggestion
	foundUsersCountMap := make(map[string]uint32, 0) // GUARDED_BY(mu).
```

similar to https://github.com/openfga/openfga/blob/1fd71eb98377280f96708d1eaf23f1e66de13a96/pkg/storage/memory/memory.go#L110

File: pkg/server/commands/listusers/list_users_rpc.go
Comment (miparnisari) [inline pkg/server/commands/listusers/list_users_rpc.go:388]:
```
                 // Compare the number of times the specific user was returned for
		// all intersection clauses plus the number of all wildcards found.
		// If this summed value >= the number of clauses, the user satisfies
		// the intersection expression and can be sent on `foundUsersChan`
		if (count + wildcardCount.Load()) >= uint32(len(clauses)) {
			foundUsersChan <- tuple.StringToUserProto(key)
		}
```

Comment (miparnisari) [inline pkg/server/commands/listusers/list_users_rpc.go:388]:
```suggestion
		if (count + wildcardCount.Load()) >= uint32(len(clauses)) {
			foundUsersChan <- tuple.StringToUserProto(key)
		}		
```

File: pkg/server/commands/listusers/list_users_rpc.go
Comment (miparnisari) [reply pkg/server/commands/listusers/list_users_rpc.go:389]: I also don't have a strong opposition if it works, so i'm not blocking the PR :) 

 I'm just concerned about the fact that we have two very different implementations of two APIs that do relatively similar things, with all the downsides that that implies (maintainability, cognitive overload etc)

File: pkg/server/commands/listusers/list_users_rpc.go
Comment (willvedd) [reply pkg/server/commands/listusers/list_users_rpc.go]: Addressed in https://github.com/openfga/openfga/pull/1490. Will pull-in that work once merged to main.

**Review by jon-whit — APPROVED** [review]

**Pull request merged**
